File: app/src/main/java/net/nekobako/twidere/extension/aniuta/MusicPlayerListenerService.java
package net.nekobako.twidere.extension.aniuta;

import android.app.Notification;
import android.content.*;
import android.os.Bundle;
import android.service.notification.*;


public class MusicPlayerListenerService extends NotificationListenerService {
	public static final String ACTION_FETCH_MUSIC_METADATA = MusicPlayerListenerService.class.getName() + ".FETCH_MUSIC_METADATA";
	private static final String NOTIFICATION_PACKAGE_NAME = "jp.co.aniuta.android.aniutaap";
	private static final int NOTIFICATION_ID = 1;
	private static final String NOTIFICATION_TITLE = "android.title";
	private static final String NOTIFICATION_ARTIST = "android.text";

	private Receiver receiver;


	@Override
	public void onCreate() {
		super.onCreate();

		this.receiver = new Receiver();
		IntentFilter intentFilter = new IntentFilter();
		intentFilter.addAction(ACTION_FETCH_MUSIC_METADATA);
		this.registerReceiver(this.receiver, intentFilter);
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		this.unregisterReceiver(this.receiver);
	}


	private class Receiver extends BroadcastReceiver {
		@Override
		public void onReceive(Context context, Intent intent) {
			for(StatusBarNotification statusBarNotification : MusicPlayerListenerService.this.getActiveNotifications()) {
				if(NOTIFICATION_PACKAGE_NAME.equals(statusBarNotification.getPackageName()) && statusBarNotification.getId() == NOTIFICATION_ID) {
					Notification notification = statusBarNotification.getNotification();
					Bundle extras = new Bundle();
					extras.putParcelable("music_metadata", new MusicMetadata((String)notification.extras.get(NOTIFICATION_TITLE), (String)notification.extras.get(NOTIFICATION_ARTIST), notification.largeIcon));
					this.setResultExtras(extras);
					return;
				}
			}
		}
	}
}

File: app/src/main/java/net/nekobako/twidere/extension/aniuta/MusicPlayerListenerService.kt
package net.nekobako.twidere.extension.aniuta

import android.content.*
import android.os.Bundle
import android.service.notification.NotificationListenerService


class MusicPlayerListenerService : NotificationListenerService() {
	companion object {
		val ACTION_FETCH_MUSIC_METADATA = "${MusicPlayerListenerService::class.java.name}.FETCH_MUSIC_METADATA"
		private val NOTIFICATION_PACKAGE_NAME = "jp.co.aniuta.android.aniutaap"
		private val NOTIFICATION_ID = 1
		private val NOTIFICATION_TITLE = "android.title"
		private val NOTIFICATION_ARTIST = "android.text"
	}


	private lateinit var receiver: Receiver


	override fun onCreate() {
		super.onCreate()

		this.receiver = Receiver()
		val intentFilter = IntentFilter()
		intentFilter.addAction(ACTION_FETCH_MUSIC_METADATA)
		this.registerReceiver(this.receiver, intentFilter)
	}

	override fun onDestroy() {
		super.onDestroy()

		this.unregisterReceiver(this.receiver)
	}


	private inner class Receiver : BroadcastReceiver() {
		override fun onReceive(context: Context?, intent: Intent?) {
			this@MusicPlayerListenerService.activeNotifications
				.singleOrNull { it.packageName == NOTIFICATION_PACKAGE_NAME && it.id == NOTIFICATION_ID }
				?.notification
				?.let { notification ->
					val extras = Bundle()
					extras.putParcelable("music_metadata", MusicMetadata(notification.extras.getString(NOTIFICATION_TITLE), notification.extras.getString(NOTIFICATION_ARTIST), notification.largeIcon))
					this.setResultExtras(extras)
				}
		}
	}
}
